cdc/generation.cc

```diff
@@ -294,6 +299,34 @@ future<db_clock::time_point> get_local_streams_timestamp() {
     });
 }
 
+// non-static for testing
+constexpr size_t limit_of_streams_in_topology_description() {
+    // Each stream takes 16B and we don't want to exceed 4MB so we can have
+    // at most 262144 streams but not less than 1 per vnode.
+    return 4 * 1024 * 1024 / 16;
+}
+
+// non-static for testing
+topology_description limit_number_of_streams_if_needed(topology_description&& desc) {
+    int64_t streams_count = 0;
+    for (auto& tr_desc : desc.entries()) {
+        streams_count += tr_desc.streams.size();
+    }
+
+    size_t limit = std::max(limit_of_streams_in_topology_description(), desc.entries().size());
+    if (limit >= streams_count) {
+        return std::move(desc);
+    }
+    size_t vnode_limit = limit / desc.entries().size();
+    auto entries = std::move(desc).entries();
+    for (auto& e : entries) {
+        if (e.streams.size() > vnode_limit) {
+            e.streams.resize(vnode_limit);
```

Comment:
Q: does this cause an imbalance somehow? Will it cause some streams to have a large number of vnodes, while others have just a few?

Reply:
One vnode can have multiple streams. Each stream always lives in a single vnode.

It is possible that some shards will get more tokens from the vnode than other shards.

I will check the exact algorithm and get back to you with more concrete answer.

Reply:
Actually no. The algorithm will think there's less shards on the machine and will load balance the token of the vnode equally among them the usual way -> https://github.com/scylladb/scylla/blob/master/cdc/metadata.cc#L40

So this will be balanced. The problem is that now token on any shard may end up on the wrong shard. Even when we remove just one stream.

Reply:
Well, that's the premise of the fix, no? That it's better to have some cross-shard traffic than those monster collections.

Reply:
In other words, for a given vnode, some shards will receive more data (that would be otherwise balanced across all shards).

However, if we look at different vnodes, then in each of them a different subset of shards will be the one receiving more data...

So maybe in the end, things will be balanced?

Reply:
Yes. I was just hoping that when there are 3 streams for 3 shards and I remove one. Then all the tokens that were mapped to stream 1 and 2 will still be mapped to those shards and only tokens mapped to stream 3 (the removed one) will end up on the wrong shards but that's not true. The algorithm will shuffle them all around.

The premise still holds though.

Reply:
Actually there will be imbalance between shards. Shards with higher numbers will get less data.

Reply:
No, there won't. Let me draw a picture...

Reply:
Maybe we could shuffle the streams before truncating the list.

Reply:
Consider the following picture:
![image](https://user-images.githubusercontent.com/6516501/107386076-cafb5c00-6af3-11eb-8fdf-357e330510a5.png)

Read slowly and the picture will become clear:
1. the black segments denote a vnode boundary. Consider this vnode, and consider its owner (some node X).
2. Suppose node X has 3 shards. Red segments denote shard boundaries for this vnode. Between each pair of segments there's a number denoting the shard that owns this segment: 0, 1, 2, 0, 1, and so on (but we don't care what happens outside the vnode for now).
3. But we trimmed the generation and created it as if there were 2 shards. So if node X had 2 shards, the blue segments denote the resulting shard boundaries: 0, 1, 0, and so on.
4. So there are two streams. We chose them somewhere in between 2 blue segments such that the "fake shard 0" owns this range - the green stream ID, and someone in between 2 blue segments such that the "fake shard 1" owns this range - the pink stream ID.
5. Partitions are mapped to stream IDs using the blue shard partitioning. Thus: all partitions in the area denoted by green go to the green stream ID, and all partitions in the area denoted by pink go to the pink stream ID.
6. In other words: shard 1 (the real one) - which owns the green stream ID - stole some data from all shards, because the green area intersects all shards. Shard 2 stole some data from shard 1 because the pink area intersects only shard 1 and shard 2.
7. Hence, within this vnode, shard 0 will get no data; the data is distributed between shard 1 and 2.

However, for each vnode, this redistribution of data will be different. If we consider all vnodes - and there's lots of them - data should be redistributed roughly equally.

Reply:
Ahh, my reasoning is incorrect here:
> We chose them somewhere in between 2 blue segments such that the "fake shard 0" owns this range - the green stream ID, and someone in between 2 blue segments such that the "fake shard 1" owns this range - the pink stream ID.

Indeed: it would be correct if we actually **calculated the stream IDs by assuming a smaller number of shards**. But currently the PR does something else: it **calculates the stream IDs assuming the original number of shards, then throws out some of them**.

So the green and pink stream IDs would lie somewhere on the first and second red range.

But maybe it's still OK; the redistribution of data will be a bit different, but it will still happen.

Reply:
Corrected picture:
![image](https://user-images.githubusercontent.com/6516501/107388348-0bf47000-6af6-11eb-8735-89e97ec590af.png)
both stream IDs would be on the "fake shard 0".

So in this vnode, shard 0 will steal all data.

Reply:
So on this picture, all data will be redistributed between shard 0 and 1.

Reply:
The data ends up on shards n, n+1, ..., (n+1) mod S, ...
where n is the shard which the left vnode boundary lies in.

Reply:
Nope, I'm wrong again.
```
        for (size_t shard_idx = 0; shard_idx < shard_count; ++shard_idx) {
            auto t = dht::find_first_token_for_shard(sharder, start, end, shard_idx);
            // compose the id from token and the "index" of the range end owning vnode
            // as defined by token sort order. Basically grouping within this
            // shard set.
            desc.streams.emplace_back(stream_id(t, index));
        }
```
The data ends up always on a prefix of shards (starting from 0), - that's how `desc.streams` is arranged.

So you're right @haaawk , the data will be imbalanced; it will always end up on a prefix of shards.

To fix it, we can either:
1. shuffle the list of streams before trimming it,
2. or calculate the stream IDs assuming a smaller number of shards in the first place.

Reply:
Perhaps a test that actually checks whether data is not "too imbalanced" would be appropriate?

Reply:
> shuffle the list of streams before trimming it,

This actually is not the best idea.
Even the current PR without shuffling is not the best idea (without even the fact that it causes shard imbalance).

The current solution may **result in assigning stream IDs from different vnodes** to data in a given vnode.
Consider this code:
```
        for (size_t shard_idx = 0; shard_idx < shard_count; ++shard_idx) {
            auto t = dht::find_first_token_for_shard(sharder, start, end, shard_idx);
            // compose the id from token and the "index" of the range end owning vnode
            // as defined by token sort order. Basically grouping within this
            // shard set.
            desc.streams.emplace_back(stream_id(t, index));
        }
```

Now, `find_first_token_for_shard` may find a token outside the vnode we're currently processing, so find a stream ID outside the vnode we're currently processing.

This is not a problem if we don't trim, because if that's the case, then it means this vnode has no tokens in this shard. So `shard_of` (in cdc/metadata.cc) would never choose this shard, so it would never choose this stream ID.

Unforutnately, with the current trimming solution, this possibility comes back: we may choose stream IDs outside the vnode (think about it).

Thus the only remaining correct solution is:
> calculate the stream IDs assuming a smaller number of shards in the first place

Reply:
`dht::find_first_token_for_shard` won't pick token from outside the range start - end

Reply:
I changed the code to use the right number of shards when generating streams ids.